core/chains/evm/gas/block_history_estimator.go

```diff
@@ -524,10 +690,36 @@ var (
 	ErrNoSuitableTransactions = errors.New("no suitable transactions")
 )
 
-func (b *BlockHistoryEstimator) percentilePrices(percentile int, eip1559 bool) (gasPrice, tipCap *big.Int, err error) {
-	gasPrices := make([]*big.Int, 0)
-	tipCaps := make([]*big.Int, 0)
-	for _, block := range b.rollingBlockHistory {
+func (b *BlockHistoryEstimator) calculatePercentilePrices(blocks []Block, percentile int, eip1559 bool, f func(gasPrices []*assets.Wei), f2 func(tipCaps []*assets.Wei)) (gasPrice, tipCap *assets.Wei, err error) {
+	gasPrices, tipCaps := b.getPercentilePricesFromBlocks(blocks, percentile, eip1559)
+	if len(gasPrices) == 0 {
+		return nil, nil, ErrNoSuitableTransactions
+	}
+	sort.Slice(gasPrices, func(i, j int) bool { return gasPrices[i].Cmp(gasPrices[j]) < 0 })
+	if f != nil {
+		f(gasPrices)
+	}
+	idx := ((len(gasPrices) - 1) * percentile) / 100
+	gasPrice = gasPrices[idx]
+
+	if eip1559 && len(tipCaps) == 0 {
+		return nil, nil, ErrNoSuitableTransactions
+	} else if eip1559 {
+		sort.Slice(tipCaps, func(i, j int) bool { return tipCaps[i].Cmp(tipCaps[j]) < 0 })
+		if f2 != nil {
+			f2(tipCaps)
+		}
+		idx := ((len(tipCaps) - 1) * percentile) / 100
+		tipCap = tipCaps[idx]
+	}
+
+	return
```

Comment:
Would it read a little easier to return early if not eip1559?
```suggestion
	if !eip1559 {
		return
	}
	if len(tipCaps) == 0 {
		return nil, nil, ErrNoSuitableTransactions
	} 
	sort.Slice(tipCaps, func(i, j int) bool { return tipCaps[i].Cmp(tipCaps[j]) < 0 })
	if f2 != nil {
		f2(tipCaps)
	}
	idx := ((len(tipCaps) - 1) * percentile) / 100
	tipCap = tipCaps[idx]
```